src/js_realm.hpp

```diff
@@ -213,6 +213,68 @@ class RealmDelegate : public BindingContext {
     friend class RealmClass<T>;
 };
 
+
+template<typename T>
+class ShouldCompactOnLaunchFunctor {
+public:
+    ShouldCompactOnLaunchFunctor(typename T::Context ctx, typename T::Function should_compact_on_launch_func)
+    : m_ctx(Context<T>::get_global_context(ctx))
+    , m_func(ctx, should_compact_on_launch_func)
+    , m_event_loop_dispatcher {ShouldCompactOnLaunchFunctor<T>::main_loop_handler}
+    , m_mutex{new std::mutex}
+    , m_cond_var{new std::condition_variable}
+    {
+    }
+
+    bool operator ()(const uint64_t total_bytes, const uint64_t used_bytes) {
+        {
+            std::lock_guard<std::mutex> lock {*m_mutex};
+            m_should_compact_on_launch_callback_done = false;
+        }
+
+        m_event_loop_dispatcher(this, total_bytes, used_bytes);
+        bool should_compact_on_launch = false;
+        {
+            std::unique_lock<std::mutex> lock(*m_mutex);
+            m_cond_var->wait(lock, [this] { return this->m_should_compact_on_launch_callback_done; });
+            should_compact_on_launch= m_should_compact_on_launch;
+        }
+
+        return should_compact_on_launch;
+    }
+
+    static void main_loop_handler(ShouldCompactOnLaunchFunctor<T>* this_object,
+                                  const uint64_t total_bytes,
+                                  const uint64_t used_bytes) {
+        const Protected<typename T::GlobalContext>& ctx = this_object->m_ctx;
+        HANDLESCOPE(ctx);
+
+        const int argc = 2;
+        typename T::Value arguments[argc] = { Value<T>::from_number(ctx, total_bytes), Value<T>::from_number(ctx, used_bytes) };
+        typename T::Value ret_val = Function<T>::callback(ctx, this_object->m_func, typename T::Object(), argc, arguments);
+        bool ret_val_bool = Value<T>::to_boolean(ctx, ret_val);
+
+        {
+            std::lock_guard<std::mutex> lock {*this_object->m_mutex};
+            this_object->m_should_compact_on_launch_callback_done = true;
+            this_object->m_should_compact_on_launch = ret_val_bool;
+        }
+
+        this_object->m_cond_var->notify_one();
+    };
+
+private:
+    const Protected<typename T::GlobalContext> m_ctx;
+    const Protected<typename T::Function> m_func;
+    util::EventLoopDispatcher<void(ShouldCompactOnLaunchFunctor<T>* this_object,
+                                   uint64_t total_bytes,
+                                   uint64_t used_bytes)> m_event_loop_dispatcher;
+    bool m_should_compact_on_launch_callback_done = false;
```

Comment:
this should not be needed if we simplify the logic. the meaning of `m_should_compact_on_launch_callback_done`  is actually  set when the sync primitive is signaled